### DISCUSSION TITLE
Custom Endpoint Permissions

### DISCUSSION THREAD

**Discussion by Thermal-Shock** [directus/directus#12078] [Archive]
Firstly, maybe I'm missing something here, but where do I set permissions on what a role/user can do with my custom endpoint?

I couldn't find anything in the docs and I would have expected to see a custom endpoints section in the admin app roles and permissions. 

I know I can access the accountability from the request, but does that mean I have to hard-code who can use it?

All the best.

**Comment by rijkvanzanten** [comment]
Heya!

> I know I can access the accountability from the request, but does that mean I have to hard-code who can use it?

That totally depends on what you're aiming to do in your endpoint. If it's data management through the provided services, simply passing the accountability object to the service constructor takes care of everything. If you have more specific needs, it's totally up to you how you'd like to handle it (you could for example add a custom field to directus_roles, and check against that when the request comes in, rather than hardcoding role IDs) 👍🏻

**Reply by Thermal-Shock** [reply]
Thanks for all your support. 

I totally get it. We could add a role, then simply check if the req.accountability is in that role. In other words, we could invent a way to have custom endpoint permissions. 

I just feel like it would be nice if when it is being registered with directus, that a new section of endpoints should appear on the roles and permissions app dashboard to control the same post, read, put/patch, and delete permissions that you can for a collection. Then within the endpoint we simply run some directus function to validate req.accountability has permission, otherwise reject it. 

Custom endpoint initial entry point permissions should be no different from entry point permissions for a collection in my mind. It represents an initial entry point to query some data using the API. The ability to control at a high level, who can call this entry point by way of built in roles and permissions, would be very convenient and save everyone having to invent a way for non-public APIs.

All the best.

**Reply by rijkvanzanten** [reply]
hmm interesting! The issue with those permissions is that it's not known ahead of time if your custom endpoint does any CRUD operations to begin with! It'd be a little difficult of Directus to render a permissions configuration UI if it can't know ahead of time what permissions the endpoint might need configured 🤔 

Another difference is that permissions don't necessarily control endpoint access. It controls what data collections you might access, but doesn't control how you access those.

**Reply by Thermal-Shock** [reply]
Certainly we are in new feature territory here, but the extension-sdk or an alternate registerEndpoints could give the programmer the ability to specify which permissions should show up in the dashboard along with a friendly name and possible other meta/presentation data. 

If the permissions did not already exist in directus for this endpoint, then it would create them and have any exposed actions as no access to match the same security philosophy as with collections. If it already exists, then it would leave the permissions as is. 

It would be up to the developer of the endpoint to decide if this endpoint is the gatekeeper to the other collections it needs to access (with or without passing accountability) when instantiating those ItemService's. 

The  data collection permissions still control direct access to the collection as normal if the user tries to go directly to it, otherwise it's up to the custom endpoint development team to decide.

EDIT: If somehow what the endpoint is allow to do with the ItemService's conflicts with data collection permissions, then the end point can also just go directly to the database (or whatever other resource) without using ItemService and authentication.  However, I assume there is no conflict here and also that Directus shouldn't really be imposing any rules on the developer decision to allow bypassing collection access through a properly registered custom endpoint. 

Thanks again for the discussion.

**Reply by rijkvanzanten** [reply]
Yeah! I can see how that'd be useful. I am wondering how that could work in practice though. A custom endpoint is fully custom logic, and could thus have any number of side-effects you might need to check permissions for (for example, you could retrieve information from Stripe, and send a text through Twilio). How would one define the permissions for those two? It'd be totally custom as compared to the existing permissions configuration in Directus currently 🤔 

I think a case like that could be solved with custom fields on the role, rather than trying to work in custom permissions, as you could add a boolean field to directus_roles to indicate `has_twilio_access` for example, which can then be used in the custom endpoint 🤔

**Reply by Thermal-Shock** [reply]
Yeah really I'm only talking about having permissions for the point of entry only, so simply CRUD or only whatever high level endpoints are registered. This would mean there was only all access or no access options in the permissions UI, and no "Use Custom". This permission would be of type "endpoint", so the UI could handle it especially or you could subclass the permissions UI for it I would presume. 

Certainly I like the idea of full flexibility and the "Use Custom" interface reading the endpoint manifest to know what permissions to show in its interface. The interface would not show the current item and field permissions, but instead pull from an endpoint manifest bundled with the extension to present whatever named access/no access controls the endpoint wanted to show. 

With all this in place, its simply a matter of having some function that can be called in the endpoint to refer to these permissions by manifest name (endpoint or deeper like twilio etc) along with the accountability and validate. 

Again, all of this allows it all to be genetically built into directus and the developer only needs to create a manifest and check those permissions as part of developing their custom endpoint bundle.
